hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala

```diff
@@ -0,0 +1,134 @@
+  def mux[A](arg: A, cond: Code[Boolean], j1: JoinPoint[A], j2: JoinPoint[A])(
+    implicit ap: ParameterPack[A]
+  ) = {
+    assert(j1.stackIndicator eq j2.stackIndicator)
+    ensureStackIndicator(j1.stackIndicator)(new Code[Ctrl] {
+      def emit(il: Growable[AbstractInsnNode]): Unit = {
+        ap.push(arg).emit(il)
+        cond.toConditional.emitConditional(il, j1.label, j2.label)
+      }
+    })
+  }
+
+  def mux(cond: Code[Boolean], j1: JoinPoint[Unit], j2: JoinPoint[Unit]): Code[Ctrl] =
+    mux((), cond, j1, j2)
+
+  case class CallCC[A: ParameterPack](
+    f: (JoinPointBuilder, JoinPoint[A]) => Code[Ctrl]
+  ) {
+    private[joinpoint] def code: Code[Nothing] = {
+      val si = new Object()
+      val jb = new JoinPointBuilder(si)
+      val ret = JoinPoint[A](new LabelNode, si)
+      val body = f(jb, ret)
+      withStackIndicator(si) {
+        Code(body, jb.define, ret.placeLabel)
+      }
+    }
+  }
+
+  /**
+    * So-called "stack-indicators" are placed on the following `mutable.Stack` during the extent of
+    * a CallCC being emitted. Whenever a join-point is called, it checks that the stack-indicator at
+    * the top of this stack is the same one associated with the CallCC from which the join-point
+    * originated. This ensures that you cannot escape a CallCC by returning from an outer CallCC,
+    * e.g.:
+    *
+    *   CallCC[Code[Int]] { (jb1, ret1) =>
+    *     ret1(const(1) + CallCC[Code[Int]] { (jb2, ret2) =>
+    *       ret1(const(2))
+    *     })
+    *   }
+    *
+    * The above code typechecks in Scala, but at runtime would corrupt the JVM stack, leaving a "1"
+    * behind without consuming it. This stack-indicator machinery catches this before bytecode
+    * verification.
+    */
+
+  class EmitLongJumpError extends AssertionError("cannot jump out of nested CallCC")
+
+  private val stack: mutable.Stack[Object] = new mutable.Stack
+
+  private def withStackIndicator[T](si: Object)(c: Code[T]) = new Code[T] {
+    def emit(il: Growable[AbstractInsnNode]): Unit = {
+      stack.push(si)
+      try
+        c.emit(il)
+      finally
+        assert(stack.pop() eq si)
+    }
+  }
+
+  private[joinpoint] def ensureStackIndicator[T](si: Object)(c: Code[T]) = new Code[T] {
+    def emit(il: Growable[AbstractInsnNode]): Unit = {
+      if (stack.top ne si) throw new EmitLongJumpError
+      c.emit(il)
+    }
+  }
+}
+
+case class JoinPoint[A] private[joinpoint](
+  label: LabelNode,
+  stackIndicator: Object
+)(implicit p: ParameterPack[A]) extends (A => Code[Ctrl]) {
+  def apply(args: A) =
+    JoinPoint.ensureStackIndicator(stackIndicator) {
+      Code(p.push(args), gotoLabel)
+    }
+
+  private[joinpoint] def placeLabel = Code(label)
+  private[joinpoint] def gotoLabel = Code[Ctrl](new JumpInsnNode(Opcodes.GOTO, label))
+}
+
+class DefinableJoinPoint[A: ParameterPack] private[joinpoint](
+  args: ParameterStore[A],
+  _stackIndicator: Object
```

Comment:
I don't think we ever use the underscore naming convention in scala, which has actual access control.